underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java

```diff
@@ -0,0 +1,486 @@
+ */
+public class S3UnderFileSystem extends UnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final String FOLDER_SUFFIX = "_$folder$";
+  private static final String PATH_SEPARATOR = "/";
+
+  private final S3Service mClient;
+  private final String mBucketName;
+
+  public S3UnderFileSystem(String bucketName, TachyonConf tachyonConf) {
+    super(tachyonConf);
+    AWSCredentials awsCredentials =
+        new AWSCredentials(tachyonConf.get(Constants.S3_ACCESS_KEY, null), tachyonConf.get(
+            Constants.S3_SECRET_KEY, null));
+    mBucketName = bucketName;
+    mClient = new RestS3Service(awsCredentials);
+  }
+
+  @Override
+  public void close() throws IOException {
+  }
+
+  @Override
+  public void connectFromMaster(TachyonConf conf, String hostname) {
+    // Authentication is taken care of in the constructor
+  }
+
+  @Override
+  public void connectFromWorker(TachyonConf conf, String hostname) {
+    // Authentication is taken care of in the constructor
+  }
+
+  @Override
+  public OutputStream create(String path) throws IOException {
+    if (mkdirs(getParentKey(path), true)) {
+      return new S3OutputStream(mBucketName, stripPrefixIfPresent(path), mClient);
+    }
+    return null;
+  }
+
+  // Same as create(path)
+  @Override
+  public OutputStream create(String path, int blockSizeByte) throws IOException {
+    LOG.warn("Create with block size is not supported with S3UnderFileSystem. Block size will be "
+        + "ignored.");
+    return create(path);
+  }
+
+  // Same as create(path)
+  @Override
+  public OutputStream create(String path, short replication, int blockSizeByte)
+      throws IOException {
+    LOG.warn("Create with block size and replication is not supported with S3UnderFileSystem."
+        + " Block size and replication will be ignored.");
+    return create(path);
+  }
+
+  @Override
+  public boolean delete(String path, boolean recursive) throws IOException {
+    if (!recursive) {
+      if (isFolder(path) && listInternal(path, false).length != 0) {
+        LOG.error("Unable to delete " + path + " because it is a non empty directory. Specify "
+            + "recursive as true in order to delete non empty directories.");
+        return false;
+      }
+      return deleteInternal(path);
+    }
+    // Get all relevant files
+    String[] pathsToDelete = listInternal(path, true);
+    for (String pathToDelete : pathsToDelete) {
+      // If we fail to deleteInternal one file, stop
+      if (!deleteInternal(CommonUtils.concatPath(path, pathToDelete))) {
+        LOG.error("Failed to delete path " + pathToDelete + ", aborting delete.");
+        return false;
+      }
+    }
+    return deleteInternal(path);
+  }
+
+  @Override
+  public boolean exists(String path) throws IOException {
+    // Root path always exists.
+    return isRoot(path) || getObjectDetails(path) != null;
+  }
+
+  // Largest size of one file is 5 TB
```

Comment:
Could you comment in javadoc why 5TB? is this a requirement from S3? or a limitation in Tachyon?